### PR TITLE
Allow a read of length 0 in InputStreamAdapter #28751

### PR DESCRIPTION
**Pull request by DarkDrek**
References #28751

I tried to write a meaningful test for this bug but I am not sure if can be written more compact.

### CODE REVIEW

**Comment by lightbend-cla-validator** [comment]
Hi @DarkDrek,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

File: akka-stream-tests/src/test/scala/akka/stream/io/InputStreamSinkSpec.scala [hunk 1]
```diff
@@ -275,6 +275,17 @@ class InputStreamSinkSpec extends StreamSpec(UnboundedMailboxConfig) {
...
+
+    "a read of length 0 should not request bytes from upstream" in assertAllStagesStopped {
+      val (probe, inputStream) = TestSource.probe[ByteString].toMat(StreamConverters.asInputStream())(Keep.both).run()
+      val f = Future(inputStream.read(new Array[Byte](byteString.size), 0, 0))
+
+      probe.executeAfterSubscription(() => probe.expectNoMessage())
+      Await.result(f, remainingOrDefault) should ===(0)
+
+      probe.sendComplete()
+      inputStream.close()
```
Comment (raboof) [inline akka-stream-tests/src/test/scala/akka/stream/io/InputStreamSinkSpec.scala]: Already looking quite good! Indeed perhaps the test could be a bit more direct, like:

```suggestion
      probe.ensureSubscription()
      probe.expectRequest()
      
      inputStream.read(new Array[Byte](byteString.size), 0, 0) should ===(0)

      probe.expectNoMessage()

      inputStream.close()
      
      probe.expectCancellation()
```

**Review by raboof — COMMENTED** [review]
One suggestion about the test, after that LGTM!

File: akka-stream-tests/src/test/scala/akka/stream/io/InputStreamSinkSpec.scala (hunk 1 above)
Comment (DarkDrek) [reply akka-stream-tests/src/test/scala/akka/stream/io/InputStreamSinkSpec.scala]: Thanks I simplified the the test

File: akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSinkStage.scala [hunk 2]
```diff
@@ -148,9 +148,13 @@ private[stream] object InputStreamSinkStage {
   override def read(a: Array[Byte], begin: Int, length: Int): Int = {
     require(a.length > 0, "array size must be >= 0")
     require(begin >= 0, "begin must be >= 0")
-    require(length > 0, "length must be > 0")
+    require(length >= 0, "length must be >= 0")
     require(begin + length <= a.length, "begin + length must be smaller or equal to the array length")
 
+    if (length == 0) {
+      return 0
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSinkStage.scala]: Return in Scala is implemented with some unpleasant trixery under the hood, better do this as an 

```
if (length == 0) 0
else {
...
}
```

Comment (DarkDrek) [reply akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSinkStage.scala]: Changed it to `if ... else ...` without braces

**Comment by johanandren** [comment]
Failure is formatting, do a compile in sbt to trigger reformat and commit/push that and it should be fixed.

**Review by johanandren — APPROVED** [review]
LGTM when the formatting is fixed

**Comment by DarkDrek** [comment]
@raboof I think your LGTM didn't get through or do you have more suggestions?

**Review by raboof — APPROVED** [review]

**Comment by raboof** [comment]
Thanks @DarkDrek!

**Pull request merged**